src/field.cpp

```diff
@@ -367,6 +369,416 @@ Option<bool> filter::parse_filter_query(const string& simple_filter_query,
+        tokens.pop();
+
+        if (isOperator(token)) {
+            // We only have two operators &&, || having the same precedence and both being left associative.
+            while (!operatorStack.empty() && operatorStack.top() != "(") {
+                postfix.push(operatorStack.top());
+                operatorStack.pop();
+            }
+
+            operatorStack.push(token);
+        } else if (token == "(") {
+            operatorStack.push(token);
+        } else if (token == ")") {
+            while (!operatorStack.empty() && operatorStack.top() != "(") {
+                postfix.push(operatorStack.top());
+                operatorStack.pop();
+            }
+
+            if (operatorStack.empty() || operatorStack.top() != "(") {
+                return Option<bool>(400, "Could not parse the filter query.");
+            }
+            operatorStack.pop();
+        } else {
+            postfix.push(token);
+        }
+    }
+
+    while(!operatorStack.empty()) {
+        if (operatorStack.top() == "(") {
+            return Option<bool>(400, "Could not parse the filter query.");
+        }
+        postfix.push(operatorStack.top());
+        operatorStack.pop();
+    }
+
+    return Option<bool>(true);
+}
+
+// https://stackoverflow.com/a/423914/11218270
+Option<bool> toParseTree(std::queue<std::string>& postfix, filter_node_t*& root) {
+    std::stack<filter_node_t*> nodeStack;
+
+    while(!postfix.empty()) {
+        auto token = postfix.front();
+        postfix.pop();
+
+        filter_node_t* filter_node;
+        if (isOperator(token)) {
+            if (nodeStack.empty()) {
+                return Option<bool>(400, "Could not parse the filter query.");
+            }
+            auto operandB = nodeStack.top();
+            nodeStack.pop();
+
+            if (nodeStack.empty()) {
+                return Option<bool>(400, "Could not parse the filter query.");
+            }
+            auto operandA = nodeStack.top();
+            nodeStack.pop();
+
+            filter_node = new filter_node_t(token, operandA, operandB);
+        } else {
+            filter_node = new filter_node_t(token);
+        }
+
+        nodeStack.push(filter_node);
+    }
+
+    if(nodeStack.empty()) {
+        return Option<bool>(400, "Could not parse the filter query.");
+    }
+
+    root = nodeStack.top();
+    return Option<bool>(true);
+}
+
+Option<bool> initFilter(filter_node_t*& node,
+                        const std::unordered_map<std::string, field>& search_schema,
+                        const Store* store,
+                        const std::string& doc_id_prefix) {
+    auto token = node->token;
+    if(!isOperator(token)) {
+        // split into [field_name, value]
+        size_t found_index = token.find(':');
+
+        if(found_index == std::string::npos) {
+            return Option<bool>(400, "Could not parse the filter query.");
+        }
+
+        std::string&& field_name = token.substr(0, found_index);
```

Comment:
`&&`?

Reply:
Didn't change the previous logic of creating a filter 
https://github.com/typesense/typesense/blob/dec1ed6533aeb04a74054020d2de603a52c334c6/src/field.cpp#L96